### PR TITLE
fix(mcp): show startup progress as each server connects

### LINKED ISSUE #5887: MCP startup can stay on “20 connecting” for a long time
During 0.9.12 dogfooding, the founder reported that the terminal continues to show a status such as “20 MCP connecting” for a long time.

The user cannot tell whether startup is making progress, a server has stalled, or the displayed count is stale.

Expected behavior:
- The connecting count settles as each server connects, fails, times out, or is cancelled.
- Long-running connection attempts expose useful status and a next action.
- Servers that have completed startup no longer count as connecting.

For v0.9.13, reproduce this with multiple configured MCP servers and distinguish real startup delays from stale UI state. Verify successful connections, failures, and stalled startup through the real terminal event loop.

Evidence: founder report only; exact duration, affected server names, and root cause have not yet been established. Do not mark this as reproduced or fixed from the report alone.

### PR DESCRIPTION
**Pull request by Hmbown**
Startup buffered every MCP connection result until the slowest batch finished, so the TUI could keep showing “20 connecting” even after individual servers were ready. Ready tools were also unavailable in the pool during that wait.

Consume each owned connection task as it completes, update the engine-owned pool under a short lock, and emit per-server progress immediately. Both startup and explicit connect-all use the same bounded task set. Preserve the eight-connection limit, catalog generations, plugin-authority validation, required-server errors, and cancellation when the task set is dropped.

Fixes #5887. The deterministic fixture proves this startup batching defect; it does not reproduce every server in the founder’s private configuration.

Validation: 157/157 focused MCP and engine nextest tests passed with the documented 16 MiB test stack; cargo check, Clippy with -D warnings, and formatting pass. A real engine event-loop fixture holds one stdio server until a release file is written and verifies that the other server is already visible and the connecting count has dropped before that release. Boot completion, retry ownership, stale-generation handling, catalog and authority tests pass.

An initial broad libtest substring filter also selected an unrelated setup test and aborted on its default-stack overflow; the focused run uses the repository’s documented stack setting. Full workspace tests were not run locally.

Based on the bundle relocation in #5890. This PR is v0.9.13 work and does not retag or republish v0.9.12.

---

> [!NOTE]
> **Medium Risk**
> Changes MCP connection lifecycle, config reload races during concurrent boot, and LLM tool-catalog/prefix pinning timing—important for correctness but covered by focused integration tests.
> 
> **Overview**
> MCP session boot no longer waits for the slowest handshake before updating the UI or tool catalog. The engine now consumes **`spawn_pending_connects`** tasks as each finishes, stores ready connections under a short pool lock, and emits **`McpSessionBoot`** progress with an updated connecting list after every completion instead of buffering until the batch ends.
> 
> Ready tools are exposed while optional servers are still connecting, and **`pending_prefix_change_reason`** is set to **`mcp-session-boot`** on progress, finished boot, and early **`mcp_tools()`** snapshots so the KV prefix refresh is scheduled even when a connection becomes ready before its mailbox event.
> 
> **`store_ready_connection`** now returns **`Result`** and rejects connections whose catalog generation no longer matches the pool (with plugin validation on store). Mid-boot config reloads call **`reload_if_config_changed`**, abort remaining connect tasks, and avoid letting stale handshakes overwrite the new configuration. Connect tasks are wrapped with panic catching via **`FutureExt`**.
> 
> New engine and pool tests cover incremental boot visibility, config invalidation during handshake, and prefix-change ordering relative to progress events.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d716c0c8aa0c45b44ed966f6d64e57c9ee5516bd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by Hmbown** [comment]
Review follow-up is pushed at 9d05c1224e. Background completion re-reads configuration, cancels obsolete batches, and refuses to re-stamp stale connection generations. Plugin authority is checked before insertion. Progress application, queue draining, and the in-flight tool snapshot all declare the catalog refresh. Handshake panics retain their server name, and Node fixtures skip cleanly when Node is unavailable.

160 focused tests pass, including ready tool availability before a stalled handshake completes, config removal/disable during that handshake, stale generation rejection, and prefix declarations. Cargo check, clippy with warnings denied, formatting, and diff checks pass. Hosted CI is running on this new head.

File: crates/tui/src/core/engine.rs [hunk 1]
```diff
@@ -6484,33 +6487,54 @@ impl Engine {
...
+                while let Some(joined) = connects.join_next().await {
+                    let (name, result) = joined
+                        .unwrap_or_else(|error| ("connection task".to_string(), Err(error.into())));
+                    remaining.retain(|pending_name| pending_name != &name);
+                    {
+                        let mut pool = pool_for_task.lock().await;
+                        // A turn may have reloaded the pool while these handshakes
+                        // were in flight. Never let their old authority or failures
+                        // overwrite the newly installed configuration.
+                        let reload = pool.reload_if_config_changed().await;
```
[the remark this answers is not in the document]
Comment (Hmbown) [reply crates/tui/src/core/engine.rs:6506]: Checked the callee: reload_if_config_changed is an async wrapper whose body only calls the synchronous reload_from_config_sources(false); it contains no await and performs no handshake/network request. Its future does not park while holding the guard. The mtime/config check is intentional at insertion so a stale handshake cannot restore revoked configuration. The existing reload implementation uses synchronous filesystem reads under the same pool authority; moving those reads off-thread would be a separate reload refactor, not a correction to the claimed suspended-await behavior. The delayed-server test verifies ready tools remain available while another handshake is stalled.

File: crates/tui/src/mcp.rs [hunk 2]
```diff
@@ -3197,20 +3199,20 @@ impl McpPool {
...
-                        errors.push((name, error));
-                    }
+            );
+            while let Some(joined) = connects.join_next().await {
+                let (name, result) = joined
+                    .unwrap_or_else(|error| ("connection task".to_string(), Err(error.into())));
+                let result = result
+                    .and_then(|connection| self.store_ready_connection(name.clone(), connection));
+                if let Err(error) = result {
+                    self.note_connect_failure(&name, &error);
```
[the remark this answers is not in the document]
Comment (Hmbown) [reply crates/tui/src/mcp.rs:3214]: In connect_all, &mut self is held for the complete batch. The catalog-generation atomic belongs to this pool (it is not a cloned Arc), and its mutation methods require borrowing this same pool; another safe caller cannot mutate it concurrently with that exclusive borrow. The config reload occurs after the batch, and that existing path clears the old pass's errors before retrying. The background engine pass is the path that releases the pool between handshakes, and it now explicitly rejects/cancels obsolete batches before recording success or failure. I do not see a reachable stale-generation/backoff case in the exclusive connect_all path.
